### PR TITLE
Local storage plugin

### PR DESCRIPTION
**Pull request by msau42**
**What this PR does / why we need it**:
Volume plugin implementation for local persistent volumes.  Scheduler predicate will direct already-bound PVCs to the node that the local PV is at.  PVC binding still happens independently.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: 
Part of #43640

**Release note**:

```
Alpha feature: Local volume plugin allows local directories to be created and consumed as a Persistent Volume.  These volumes have node affinity and pods will only be scheduled to the node that the volume is at.
```

### CODE REVIEW

**Comment by msau42** [comment]
/assign @saad-ali @jsafrane 

cc @kubernetes/sig-storage-pr-reviews

File: pkg/volume/local_volume/local_volume.go [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+		v1.ReadWriteOnce,
+	}
+}
+
+func getVolumeSource(spec *volume.Spec) (*v1.LocalVolumeSource, bool, error) {
+	if spec.PersistentVolume != nil && spec.PersistentVolume.Spec.LocalVolume != nil {
+		return spec.PersistentVolume.Spec.LocalVolume, spec.ReadOnly, nil
+	}
+
+	return nil, false, fmt.Errorf("Spec does not reference an LocalVolume volume type")
```
Comment (rootfs) [inline pkg/volume/local_volume/local_volume.go]: an -> a

File: pkg/volume/local_volume/local_volume.go [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+				return err
+			}
+		}
+		os.Remove(dir)
+		glog.Errorf("Mount of volume %s failed: %v", dir, err)
+		return err
+	}
+
+	// TODO: how to prevent multiple mounts with conflicting fsGroup?
+	return volume.SetVolumeOwnership(m, fsGroup)
```
Comment (rootfs) [inline pkg/volume/local_volume/local_volume.go]:
```go
if ! m.readOnly {
  return volume.SetVolumeOwnership(m, fsGroup)
}
```

File: pkg/volume/local_volume/local_volume.go [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+	*localVolume
+	readOnly bool
+}
+
+var _ volume.Mounter = &localVolumeMounter{}
+
+func (m *localVolumeMounter) GetAttributes() volume.Attributes {
+	return volume.Attributes{
+		ReadOnly:        m.readOnly,
+		Managed:         false,
```
Comment (rootfs) [inline pkg/volume/local_volume/local_volume.go]:
```go
Managed:   !m.readOnly
```

File: pkg/volume/local_volume/local_volume.go [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+		SupportsSELinux: false,
```
Comment (rootfs) [inline pkg/volume/local_volume/local_volume.go]: why false?

File: pkg/volume/local_volume/local_volume.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+	}
+}
+
+// Checks prior to mount operations to verify that the required components (binaries, etc.)
```
Comment (rootfs) [inline pkg/volume/local_volume/local_volume.go]: `CanMount checks ...`

File: pkg/volume/local_volume/local_volume.go [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+	hostNode := plugin.host.GetNodeName()
+	if volumeSource.NodeName != hostNode {
+		return nil, fmt.Errorf("LocalVolume NodeName %q does not match this node %q", volumeSource.NodeName, hostNode)
+	}
+
+	return &localVolumeMounter{
+		localVolume: &localVolume{
+			podUID:     pod.UID,
+			volName:    spec.Name(),
+			mounter:    plugin.host.GetMounter(),
```
Comment (rootfs) [inline pkg/volume/local_volume/local_volume.go]: use `SafeFormatAndMount` to format the disk if not formatted?

Comment (msau42) [inline pkg/volume/local_volume/local_volume.go]: For now, the volume specified has to already have a filesystem.  We will add block support in the future.

File: pkg/volume/local_volume/local_volume.go (hunk 4 above)
Comment (msau42) [inline pkg/volume/local_volume/local_volume.go]: I don't have cycles to test and verify this.  Anyone else is welcome to take this up :-)

File: pkg/volume/local_volume/local_volume.go (hunk 3 above)
Comment (msau42) [inline pkg/volume/local_volume/local_volume.go]: What does Managed mean?  It's not documented...

File: pkg/volume/local_volume/local_volume.go [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+func (plugin *localVolumePlugin) SupportsMountOption() bool {
+	return false
+}
+
+func (plugin *localVolumePlugin) SupportsBulkVolumeVerification() bool {
+	return false
+}
+
+func (plugin *localVolumePlugin) GetAccessModes() []v1.PersistentVolumeAccessMode {
+	// The current meaning of AccessMode is how many nodes can attach to it, not how many pods can mount it
```
Comment (msau42) [inline pkg/volume/local_volume/local_volume.go]: This is the way that access modes work in the code right now, contrary to what most people assume.  It's an issue with all volumes, not just local volumes.  But perhaps local volumes will reveal this discrepancy more than other types.

I think there are two approaches to fix this: 1) Change the semantic of current AccessMode, and hopefully nobody was relying on the old behavior.  2) Add a new kind of AccessMode to enforce at the pod level.

Comment (ddysher) [inline pkg/volume/local_volume/local_volume.go]: I can't tell which approach is better at this moment.  But from a user's perspective, this is definitely a problem if there is such a big misunderstanding.  I'll see if I can help when I get there.

File: pkg/volume/local_volume/local_volume.go [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+		localVolume: &localVolume{
+			podUID:  podUID,
+			volName: volName,
+			mounter: plugin.host.GetMounter(),
+			plugin:  plugin,
+		},
+	}, nil
+}
+
+func (plugin *localVolumePlugin) ConstructVolumeSpec(volumeName, mountPath string) (*volume.Spec, error) {
```
Comment (msau42) [inline pkg/volume/local_volume/local_volume.go]: @chakri-nelluri Can you check if this is correct?  This volume can only be specified as a PV, so I am using the PV name as the volume's name.

File: pkg/volume/local_volume/local_volume.go (hunk 4 above)
Comment (ddysher) [inline pkg/volume/local_volume/local_volume.go]: In SELinux enabled cluster, we should have this set to true in order to share local volumes even for containers in a pod.  This is also required to share volume among different pods.

We already have user manage and provide SELinux labels right now, so setting this to true should suffice.

also commented here https://github.com/kubernetes/kubernetes/issues/45054

File: pkg/volume/local_volume/local_volume.go (hunk 8 above)
Comment (chakri-nelluri) [inline pkg/volume/local_volume/local_volume.go]: Hi, it looks good to me. Since PV Name is unique, it can be used.

**Comment by msau42** [comment]
API is in, so I've rebased this PR to pick up those changes.

File: test/e2e/storage/persistent_volumes-local.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+	}
+
+	By("Cleaning up PVC and PV")
+	errs := framework.PVPVCCleanup(config.client, config.ns, volume.pv, volume.pvc)
+	if len(errs) > 0 {
+		framework.Logf("AfterEach: Failed to delete PV and/or PVC: %v", utilerrors.NewAggregate(errs))
+	}
+
+	By("Removing the test directory")
+	writeCmd := fmt.Sprintf("rm -r %s", volume.containerDir)
```
Comment (mikegrass) [inline test/e2e/storage/persistent_volumes-local.go]: Nit: maybe rename to removeCmd to match the intent?

**Review by mikegrass — COMMENTED** [review]
Minor nits.

**Comment by msau42** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by msau42** [comment]
@k8s-bot pull-kubernetes-node-e2e test this

**Comment by msau42** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by msau42** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by msau42** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by sky-big** [comment]
Hi, When I use the local presistent plugin, have one question, this is my yaml configuration:  

apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    volume.alpha.kubernetes.io/node-affinity:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: kubernetes.io/hostname
              operator: In
              values:
              - 192.168.152.134
  name: my-pv
spec:
  capacity:
    storage: 5Gi
  accessModes:
    - ReadWriteOnce
  local:
    path: /tmp/k8s-storage

This configuration can't run success, the error main reason is nodeAffinity, when I delete nodeAffinity and delete validation.go file about nodeAffinity check code, the configuration is :

apiVersion: v1
kind: PersistentVolume
metadata:
  name: my-pv
spec:
  capacity:
    storage: 5Gi
  accessModes:
    - ReadWriteOnce
  local:
    path: /tmp/k8s-storage

then success. So I want to ask how to configure the local volume plugin nodeAffinity ?

**Comment by msau42** [comment]
@sky-big Try the example [here](https://github.com/kubernetes-incubator/external-storage/blob/master/local-volume/examples/manual-pv.yaml)

**Comment by sky-big** [comment]
thanks @msau42

**Comment by groundnuty** [comment]
@msau42 I've having a success with the example you linked above and I've integrated with simple nginx log persistence:

~~~
kind: Pod
apiVersion: v1
metadata:
  name: task-pv-pod
spec:
  volumes:
    - name: task-pv-storage
      persistentVolumeClaim:
       claimName: example-local-claim
  containers:
    - name: task-pv-container
      image: nginx
      ports:
        - containerPort: 80
          name: "http-server"
      volumeMounts:
      - mountPath: "/var/log/nginx/"
        name: task-pv-storage
---
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: example-local-claim
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: local-storage
---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: task-pv-volume
  annotations:
    "volume.alpha.kubernetes.io/node-affinity": '{
        "requiredDuringSchedulingIgnoredDuringExecution": {
            "nodeSelectorTerms": [
                { "matchExpressions": [
                    { "key": "kubernetes.io/hostname",
                      "operator": "In",
                      "values": ["k8s-cyfronet-1-worker10"]
                    }
                ]}
             ]}
          }'
spec:
  capacity:
    storage: 10Gi
  accessModes:
    - ReadWriteOnce
  persistentVolumeReclaimPolicy: Delete
  storageClassName: local-storage
  local:
    path: /mnt/storage/k8s_local_volumes/
~~~

The small issue is `persistentVolumeReclaimPolicy`: when set to Delete and deleting the pv/task-pvc-volume

The PV enters the `Failed` state with a message:
~~~
Error getting deleter volume plugin for volume "task-pv-volume": no deletable volume plugin matched
~~~

The desired behavior would be `rm -rf  /mnt/storage/k8s_local_volumes/*`. 

Am I doing something wrong, or is normal in alpha?

**Comment by msau42** [comment]
@groundnuty currently the volume deletion is handled by the [external provisioner](https://github.com/kubernetes-incubator/external-storage/tree/master/local-volume), so if you're not using that, then the cleanup will fail.  The idea was that you may want to customize your deletion method (zero out the volume, write random bytes x times, different methods for hdds vs ssds), so you could configure the provisioner to do whatever you wanted.  But for the case where you're not using the external provisioner, I can look into providing a default deletion routine that just removes the contents.

**Comment by msau42** [comment]
@groundnuty sorry I remember the challenge now.  Delete is called by the master controller and not the kubelet, so we are not able to access the local volume to do any cleanup.  That is why we had to implement it in the external provisioner, which runs as a daemonset on every node.

**Comment by groundnuty** [comment]
@msau42 the idea is very good indeed :) tho in the simplest usecase this feature can work just like `hostpath` with smart scheduling of a pod to a place where the volume is. Then the same simple reclaim/delete strategies that apply to `hostpath` and `nfs` would apply here. 

also with a simple setup like this (no external prov.):
~~~
kind: Pod
apiVersion: v1
metadata:
  name: task-pv-pod
spec:
  volumes:
    - name: task-pv-storage
      persistentVolumeClaim:
       claimName: example-local-claim
  containers:
    - name: task-pv-container
      image: nginx
      ports:
        - containerPort: 80
          name: "http-server"
      volumeMounts:
      - mountPath: "/var/log/nginx/"
        name: task-pv-storage
---
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: example-local-claim
spec:
  accessModes:
  - ReadWriteMany
  resources:
    requests:
      storage: 10Gi
  storageClassName: local-storage
---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: task-pv-volume
  annotations:
    "volume.alpha.kubernetes.io/node-affinity": '{
        "requiredDuringSchedulingIgnoredDuringExecution": {
            "nodeSelectorTerms": [
                { "matchExpressions": [
                    { "key": "beta.kubernetes.io/os",
                      "operator": "In",
                      "values": ["linux"]
                    }
                ]}
             ]}
          }'
spec:
  capacity:
    storage: 10Gi
  accessModes:
    - ReadWriteMany
  persistentVolumeReclaimPolicy: Delete
  storageClassName: local-storage
  local:
    path: /mnt/storage/k8s_local_volumes/
~~~

The command ` kubectl describe pvc/example-local-claim` says:
~~~
persistentvolume-controller                     Warning         ProvisioningFailed      storageclass.storage.k8s.io "local-storage" not found
~~~
as a consequence (I suppose) the nginx pod is stuck in a `ContainerCreating` state for exceptionally long time (with the image present on the node) then it fails, get recreated and all is fine, detailed log [here](https://nopaste.me/view/raw/87dd5e35).

**Comment by groundnuty** [comment]
@msau42 regarding the challenge, if you had time I would be grateful if you could outline what steps one should undertake to make the above use case work

**Comment by msau42** [comment]
The simplest way is to use the external provisioner daemonset.

Changing the pv controller to call Delete() on the nodes instead of the master is a big architectural change that only makes sense for local volumes, and no other volume plugin.

**Comment by groundnuty** [comment]
@msau42 The provisioner is working nicely: I've created one dir on a node and it created a volume for it.

I've trying to use your stateful set example with replica set to 1 (the volumecalimtemplate is convinient), unfortunately despite volume claim, claiming it is bound to a proper volume and volume claiming in turns it's used by a volume claim. I get errors in a stateful set pod that claims that a volume is not bound and it results in a pod being scheduled to a wrong node:
https://nopaste.me/view/834b80d7

Any ideas?

**Pull request merged**